### PR TITLE
[rush-lib] Support PNPM injected install feature for Rush Subspace

### PR DESCRIPTION
**Pull request by g-chao**
## Summary
This PR is to support [PNPM injected installation](https://pnpm.io/package_json#dependenciesmeta) in Rush subspace feature

## Details
Let's say we have two subspaces (S1, S2) in a Rush Monorepo, and set up is like this:
- subspace S1 / lockfile L1 contains project A and B
- subspace S2 / lockfile L2 contains project C and D
- A has an injected workspace:* dependency on C
- B has a non-injected workspace:* dependency on C
- C has a non-injected workspace:* dependency on D

Here, we have two cases:
1. A has an injected workspace:* dependency on C, the dependency graph is like A -> (injected) C ->(no-injected) D
In this case, we think If a package gets installed in the injected way then every workspace:* dependency of that package also gets injected.
Why? Recall the injected install definition from [PNPM](https://pnpm.io/package_json#dependenciesmeta), If we install a local package in the injected way, that package will be hard linked to the virtual store (node_modules/.pnpm) and symlinked from the virtual store to the modules directory. It is like we are installing that package from a NPM registry. And a package from NPM can not contain any local dependencies. 

2. In our example, A -> (injected) C, while B -> (no-injected) C, and A and B are in the same subspace, so will they conflict with each other? 
No, every package will install its dependencies in its own node_modules folder. In our case, inside A's node_modules folder, C is hard linked to virtual store (node_modules/.pnpm), inside B's node_modules, C is symlinked directly from its location in the workspace. 

Please note: If there is a case that all projects inside a dependency graph are in same subspace (no cross subspace dependencies), then we do nothing, let PNPM handle it. 

## Implementation
After we understand the issue, the implementation is fairly straightforward. 
1. We can analyze the dependency graph when generating the `pnpmfileSettings.json` file, put all necessary info we need to decide if a workspace dependency should installed in the injected way or no-injected way. 
2. Inside PNPM's `readPackage` hook, rewrite the workspace dependency installation behavior based on`pnpmfileSettings.json` file.
## How it was tested
Manually tested by setting two subspaces in Rushstack repo. 
After this PR merged, a follow up PR will enable subspace feature for Rushstack repo, so we will have more tests in that PR.

### CODE REVIEW

**Comment by iclanton** [comment]
Would you fill out the PR description, please?

**Comment by g-chao** [comment]
> Would you fill out the PR description, please?

I will do it later, right now, I am just utilize the CI to do some verification

File: libraries/rush-lib/src/logic/pnpm/SubspacePnpmfileConfiguration.ts [hunk 1]
```diff
@@ -86,4 +94,92 @@ export class SubspacePnpmfileConfiguration {
...
+
+      if (currentProject && devDependencies) {
+        SubspacePnpmfileConfiguration._processDependenciesForTransitiveInjectedInstall(
+          projectNameToInjectedDependenciesMap,
+          processTransitiveInjectedInstallQueue,
+          devDependencies,
+          currentProject,
+          rushConfiguration
+        );
+      }
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/pnpm/SubspacePnpmfileConfiguration.ts:161]:
```suggestion
      if (currentProject) {
        if (dependencies) {
          SubspacePnpmfileConfiguration._processDependenciesForTransitiveInjectedInstall(
            projectNameToInjectedDependenciesMap,
            processTransitiveInjectedInstallQueue,
            dependencies,
            currentProject,
            rushConfiguration
          );
        }

        if (devDependencies) {
          SubspacePnpmfileConfiguration._processDependenciesForTransitiveInjectedInstall(
            projectNameToInjectedDependenciesMap,
            processTransitiveInjectedInstallQueue,
            devDependencies,
            currentProject,
            rushConfiguration
          );
        }
      }
```

File: libraries/rush-lib/src/logic/pnpm/SubspacePnpmfileConfiguration.ts [hunk 2]
```diff
@@ -86,4 +94,92 @@ export class SubspacePnpmfileConfiguration {
...
+    }
+
+    return projectNameToInjectedDependenciesMap;
+  }
+
+  private static _processDependenciesForTransitiveInjectedInstall(
+    projectNameToInjectedDependencies: Map<string, Set<string>>,
+    processTransitiveInjectedInstallQueue: Array<RushConfigurationProject> = [],
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/pnpm/SubspacePnpmfileConfiguration.ts:169]: When wouldn't this be provided?

Comment (g-chao) [reply libraries/rush-lib/src/logic/pnpm/SubspacePnpmfileConfiguration.ts:171]: Sorry, I did not get the point here. What is not provided here?

Comment (octogonz) [reply libraries/rush-lib/src/logic/pnpm/SubspacePnpmfileConfiguration.ts:171]: I think Ian's point is that this parameter doesn't need to be optional.  Also, optional parameters generally should be avoided except in public APIs with heavy usage.  For example:

```ts
function doOtherThing(url: string, workingFolder?: string): void {
}

function doThing(url: string, workingFolder?: string): void {
  doOtherThing(url); // <- mistake
}
```

This mistake won't happen with `string | undefined`.

**Pull request merged**